app/services/preserved_object_handler.rb

```diff
@@ -0,0 +1,159 @@
+    DB_UPDATE_FAILED => "db update failed: %{addl}"
+  }.freeze
+
+  include ActiveModel::Validations
+
+  # Note: supplying validations here to allow validation before use, e.g. incoming_version in numeric logic
+  validates :druid, presence: true, format: { with: /\A[a-z]{2}[0-9]{3}[a-z]{2}[0-9]{4}\z/ }
+  validates :incoming_version, presence: true, numericality: { only_integer: true, greater_than: 0 }
+  validates :incoming_size, numericality: { only_integer: true, greater_than: 0 }
+
+  attr_reader :druid, :incoming_version, :incoming_size
+
+  def initialize(druid, incoming_version, incoming_size)
+    @druid = druid
+    @incoming_version = version_string_to_int(incoming_version)
+    @incoming_size = string_to_int(incoming_size)
+  end
+
+  def update_or_create
+    results = []
+
+    if invalid?
+      results << result_hash(INVALID_ARGUMENTS, errors.full_messages)
+    elsif PreservedObject.exists?(druid: druid)
+      Rails.logger.debug "update #{druid} called and object exists"
+
+      db_object = PreservedObject.find_by(druid: druid)
+      results << update_per_version_comparsion(db_object)
+    else
+      pp_default = PreservationPolicy.create(preservation_policy_name: 'default',
+                                             archive_ttl: 604_800,
+                                             fixity_ttl: 604_800)
+      PreservedObject.create(druid: druid,
+                             current_version: incoming_version,
+                             size: incoming_size,
+                             preservation_policy: pp_default)
+      results << result_hash(CREATED_NEW_OBJECT)
+    end
+    results.flatten!
+    log_results(results)
+    results
+  end
+
+  private
+
+  # expects @incoming_version to be numeric
+  # TODO: update existence check timestamps/status per each flavor of comparison?
+  def update_per_version_comparsion(db_object)
+    version_comparison = db_object.current_version <=> incoming_version
+    results = []
+    if version_comparison.zero?
+      results << result_hash(VERSION_MATCHES)
+    elsif version_comparison == 1
+      # TODO: needs manual intervention until automatic recovery services implemented
+      results << result_hash(ARG_VERSION_LESS_THAN_DB_OBJECT)
+    elsif version_comparison == -1
+      db_object.current_version = incoming_version
+      db_object.size = incoming_size if incoming_size
+      results << result_hash(ARG_VERSION_GREATER_THAN_DB_OBJECT)
+    end
+
+    update_db_object(db_object, results)
+    results.flatten
+  end
+
+  # TODO: this may need reworking if we need to distinguish db timestamp updates when
+  #   version matched vs. incoming version less than db object
+  def update_db_object(db_object, results)
+    if db_object.changed?
+      db_object.save
+      results << result_hash(UPDATED_DB_OBJECT)
+    else
+      # FIXME: we may not want to do this, but instead to update specific timestamp for check
+      db_object.touch
```

Comment:
Created #141

Reply:
commented on #141;  assume this isn't a blocker for merging.